src/firmware/app/primitives/direct_control_primitive.c

```diff
@@ -0,0 +1,141 @@
+DEFINE_PRIMITIVE_STATE_CREATE_AND_DESTROY_FUNCTIONS(DirectControlPrimitiveState_t)
+
+void app_direct_control_primitive_start(DirectControlPrimitiveMsg prim_msg,
+                                        void* void_state_ptr, FirmwareWorld_t* world)
+{
+    DirectControlPrimitiveState_t* state = (DirectControlPrimitiveState_t*)void_state_ptr;
+    const FirmwareRobot_t* robot         = app_firmware_world_getRobot(world);
+    Chicker_t* chicker                   = app_firmware_robot_getChicker(robot);
+    Dribbler_t* dribbler                 = app_firmware_robot_getDribbler(robot);
+    Charger_t* charger                   = app_firmware_robot_getCharger(robot);
+
+    switch (prim_msg.which_wheel_control)
+    {
+        case DirectControlPrimitiveMsg_direct_per_wheel_control_tag:
+        {
+            DirectControlPrimitiveMsg_DirectPerWheelControlMsg control_msg =
+                prim_msg.wheel_control.direct_per_wheel_control;
+            state->direct_velocity = false;
+            // TODO (#1649): Fix passing rpm into an applyForce function
+            app_wheel_applyForce(app_firmware_robot_getFrontLeftWheel(robot),
+                                 control_msg.front_left_wheel_rpm);
+            app_wheel_applyForce(app_firmware_robot_getBackLeftWheel(robot),
+                                 control_msg.back_left_wheel_rpm);
+            app_wheel_applyForce(app_firmware_robot_getBackRightWheel(robot),
+                                 control_msg.back_right_wheel_rpm);
+            app_wheel_applyForce(app_firmware_robot_getFrontRightWheel(robot),
+                                 control_msg.front_right_wheel_rpm);
+            break;
+        }
+        case DirectControlPrimitiveMsg_direct_velocity_control_tag:
+        {
+            DirectControlPrimitiveMsg_DirectVelocityControlMsg control_msg =
+                prim_msg.wheel_control.direct_velocity_control;
+            state->direct_velocity          = true;
+            state->direct_target_velocity_x = control_msg.velocity.x_component_meters;
+            state->direct_target_velocity_y = control_msg.velocity.y_component_meters;
+            state->direct_target_velocity_angular =
+                control_msg.angular_velocity.radians_per_second;
+            break;
+        }
+        default:
+        {
+            // Do nothing
+            // TODO (Issue #1652): LOG(WARNING)<<"Wheel control command is not a valid
+            // type"
+            state->direct_velocity = false;
+        }
+    }
+
+    switch (prim_msg.charge_mode)
+    {
+        case DirectControlPrimitiveMsg_ChargeMode_CHARGE:
+        {
+            app_charger_charge_capacitor(charger);
+            break;
+        }
+        case DirectControlPrimitiveMsg_ChargeMode_FLOAT:
+        {
+            app_charger_float_capacitor(charger);
+            break;
+        }
+        default:
+        {
+            // discharge
+            app_chicker_disableAutochip(chicker);
```

Comment:
Bit of a meta-question here. Should discharging the robot imply disabling autokick?

In most cases this does make sense since the robot can't actually actuate the chicker without being charged so leaving autokick enabled while discharged results in basically the same behavior. However the one case that's specific to RobotDiagnostics I can think of is that the breakbeam is connected to an LED, so if we want to stick something in the breakbeam to check it's working would be safer while discharged.

Edit: I realized that's not a great example since the breakbeam is connected to the LED regardless of what chick mode the robot is in. Still curious what people think, but this isn't blocking.

Reply:
I think `float` would cover a situation where capacitor could be largely discharged, but we can have autokick on?

Reply:
IMO enabling/disabling auto{kick|chip} should charge/discharge respectively, but charging/discharging should not have any effect on auto{chip|kick} being enabled or disabled.

Reply:
@garethellis0 I disagree with autokick affecting charge. We don't want to constantly be charging/discharging if we switch between primitives that do/don't have autokick, and it would also mean we can't kick for about 0.5-1 second when starting a kicking primitive because the caps need time to charge.

Reply:
> I think float would cover a situation where capacitor could be discharged, but we can have autokick on?

I'm not sure that's necessarily better than just leaving the caps charged though? Maybe a question for elec as to what happens electrically when we charge/float/discharge, since the only concrete argument I'd have against floating is if changing from charge/float constantly was somehow bad electrically.

Reply:
changed charger discharge to not disable autokick/autochip. Changing autokick and autochip behaviour is out of scope of this PR.